File: frontend/src/lib/api/macsec.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface MacsecMkaConfig {
  cak: string | null;
  ckn: string | null;
  priority: string | null;
}

export interface MacsecStaticPeer {
  name: string;
  key: string | null;
  mac: string | null;
  disable: boolean;
}

export interface MacsecStaticConfig {
  key: string | null;
  peers: MacsecStaticPeer[];
}

export interface MacsecSecurityConfig {
  cipher: string | null;
  encrypt: boolean;
  replay_window: string | null;
  mka: MacsecMkaConfig | null;
  static: MacsecStaticConfig | null;
}

export interface MacsecIpSettings {
  adjust_mss: string | null;
  arp_cache_timeout: string | null;
  disable_arp_filter: boolean;
  disable_forwarding: boolean;
  enable_arp_accept: boolean;
  enable_arp_announce: boolean;
  enable_arp_ignore: boolean;
  enable_directed_broadcast: boolean;
  enable_proxy_arp: boolean;
  proxy_arp_pvlan: boolean;
  source_validation: string | null;
}

export interface MacsecIpv6Settings {
  accept_dad: string | null;
  address_autoconf: boolean;
  address_eui64: string | null;
  address_no_default_link_local: boolean;
  address_interface_identifier: string | null;
  adjust_mss: string | null;
  base_reachable_time: string | null;
  disable_forwarding: boolean;
  dup_addr_detect_transmits: string | null;
  source_validation: string | null;
}

export interface MacsecDhcpOptions {
  client_id: string | null;
  default_route_distance: string | null;
  host_name: string | null;
  mtu: boolean;
  no_default_route: boolean;
  reject: string | null;
  user_class: string | null;
  vendor_class_id: string | null;
}

export interface MacsecDhcpv6PdInterface {
  address: string | null;
  sla_id: string | null;
}

export interface MacsecDhcpv6Options {
  duid: string | null;
  no_release: boolean;
  no_request_dns: boolean;
  no_request_domain_name: boolean;
  parameters_only: boolean;
  rapid_commit: boolean;
  temporary: boolean;
  pd: Record<string, { length: string | null; interfaces: Record<string, MacsecDhcpv6PdInterface> }> | null;
}

export interface MacsecInterface {
  name: string;
  type: string;
  addresses: string[];
  description: string | null;
  disabled: boolean;
  mtu: string | null;
  source_interface: string | null;
  vrf: string | null;
  security: MacsecSecurityConfig | null;
  ip: MacsecIpSettings | null;
  ipv6: MacsecIpv6Settings | null;
  dhcp_options: MacsecDhcpOptions | null;
  dhcpv6_options: MacsecDhcpv6Options | null;
  mirror_ingress: string | null;
  mirror_egress: string | null;
  redirect: string | null;
}

export interface MacsecConfigResponse {
  interfaces: MacsecInterface[];
  total: number;
}

export interface MacsecCapabilities {
  version: string;
  features: Record<string, { supported: boolean; description: string }>;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

export interface MacsecBatchOperation {
  op: string;
  value?: string;
}

// ============================================================================
// API Service
// ============================================================================

class MacsecService {
  async getCapabilities(): Promise<MacsecCapabilities> {
    return apiClient.get<MacsecCapabilities>("/vyos/macsec/capabilities");
  }

  async getConfig(refresh: boolean = false): Promise<MacsecConfigResponse> {
    return apiClient.get<MacsecConfigResponse>("/vyos/macsec/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post("/vyos/config/refresh");
  }

  async batchConfigure(
    interfaceName: string,
    operations: MacsecBatchOperation[]
  ): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/macsec/batch", {
      interface: interfaceName,
      operations,
    });
    await this.refreshConfig();
    return result;
  }

  async createInterface(config: {
    name: string;
    source_interface: string;
    description?: string;
    mtu?: string;
    vrf?: string;
    disabled?: boolean;
    addresses?: string[];
    security?: {
      cipher?: string;
      encrypt?: boolean;
      replay_window?: string;
      mka?: { cak?: string; ckn?: string; priority?: string };
      static_key?: string;
      static_peers?: { name: string; key?: string; mac?: string; disable?: boolean }[];
    };
    ip?: {
      adjust_mss?: string;
      arp_cache_timeout?: string;
      disable_arp_filter?: boolean;
      disable_forwarding?: boolean;
      enable_arp_accept?: boolean;
      enable_arp_announce?: boolean;
      enable_arp_ignore?: boolean;
      enable_directed_broadcast?: boolean;
      enable_proxy_arp?: boolean;
      proxy_arp_pvlan?: boolean;
      source_validation?: string;
    };
    ipv6?: {
      accept_dad?: string;
      address_autoconf?: boolean;
      address_eui64?: string;
      address_no_default_link_local?: boolean;
      address_interface_identifier?: string;
      adjust_mss?: string;
      base_reachable_time?: string;
      disable_forwarding?: boolean;
      dup_addr_detect_transmits?: string;
      source_validation?: string;
    };
    dhcp_options?: {
      client_id?: string;
      default_route_distance?: string;
      host_name?: string;
      mtu?: boolean;
      no_default_route?: boolean;
      reject?: string;
      user_class?: string;
      vendor_class_id?: string;
    };
    dhcpv6_options?: {
      duid?: string;
      no_release?: boolean;
      parameters_only?: boolean;
      rapid_commit?: boolean;
      temporary?: boolean;
      no_request_dns?: boolean;
      no_request_domain_name?: boolean;
    };
    mirror_ingress?: string;
    mirror_egress?: string;
    redirect?: string;
  }): Promise<VyOSResponse> {
    const operations: MacsecBatchOperation[] = [];

    // Source interface is required
    operations.push({ op: "set_source_interface", value: config.source_interface });

    if (config.description) operations.push({ op: "set_interface_description", value: config.description });
    if (config.mtu) operations.push({ op: "set_mtu", value: config.mtu });
    if (config.vrf) operations.push({ op: "set_vrf", value: config.vrf });
    if (config.disabled) operations.push({ op: "set_interface_disable" });

    // Addresses
    if (config.addresses) {
      for (const addr of config.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }

    // Security
    if (config.security) {
      if (config.security.cipher) operations.push({ op: "set_security_cipher", value: config.security.cipher });
      if (config.security.encrypt) operations.push({ op: "set_security_encrypt" });
      if (config.security.replay_window) operations.push({ op: "set_security_replay_window", value: config.security.replay_window });

      // MKA
      if (config.security.mka) {
        if (config.security.mka.cak) operations.push({ op: "set_security_mka_cak", value: config.security.mka.cak });
        if (config.security.mka.ckn) operations.push({ op: "set_security_mka_ckn", value: config.security.mka.ckn });
        if (config.security.mka.priority) operations.push({ op: "set_security_mka_priority", value: config.security.mka.priority });
      }

      // Static
      if (config.security.static_key) operations.push({ op: "set_security_static_key", value: config.security.static_key });
      if (config.security.static_peers) {
        for (const peer of config.security.static_peers) {
          operations.push({ op: "set_security_static_peer", value: peer.name });
          if (peer.key) operations.push({ op: "set_security_static_peer_key", value: `${peer.name}:${peer.key}` });
          if (peer.mac) operations.push({ op: "set_security_static_peer_mac", value: `${peer.name}:${peer.mac}` });
          if (peer.disable) operations.push({ op: "set_security_static_peer_disable", value: peer.name });
        }
      }
    }

    // IP settings
    if (config.ip) {
      if (config.ip.adjust_mss) operations.push({ op: "set_ip_adjust_mss", value: config.ip.adjust_mss });
      if (config.ip.arp_cache_timeout) operations.push({ op: "set_ip_arp_cache_timeout", value: config.ip.arp_cache_timeout });
      if (config.ip.disable_arp_filter) operations.push({ op: "set_ip_disable_arp_filter" });
      if (config.ip.disable_forwarding) operations.push({ op: "set_ip_disable_forwarding" });
      if (config.ip.enable_arp_accept) operations.push({ op: "set_ip_enable_arp_accept" });
      if (config.ip.enable_arp_announce) operations.push({ op: "set_ip_enable_arp_announce" });
      if (config.ip.enable_arp_ignore) operations.push({ op: "set_ip_enable_arp_ignore" });
      if (config.ip.enable_directed_broadcast) operations.push({ op: "set_ip_enable_directed_broadcast" });
      if (config.ip.enable_proxy_arp) operations.push({ op: "set_ip_enable_proxy_arp" });
      if (config.ip.proxy_arp_pvlan) operations.push({ op: "set_ip_proxy_arp_pvlan" });
      if (config.ip.source_validation) operations.push({ op: "set_ip_source_validation", value: config.ip.source_validation });
    }

    // IPv6 settings
    if (config.ipv6) {
      if (config.ipv6.accept_dad) operations.push({ op: "set_ipv6_accept_dad", value: config.ipv6.accept_dad });
      if (config.ipv6.address_autoconf) operations.push({ op: "set_ipv6_address_autoconf" });
      if (config.ipv6.address_eui64) operations.push({ op: "set_ipv6_address_eui64", value: config.ipv6.address_eui64 });
      if (config.ipv6.address_no_default_link_local) operations.push({ op: "set_ipv6_address_no_default_link_local" });
      if (config.ipv6.address_interface_identifier) operations.push({ op: "set_ipv6_address_interface_identifier", value: config.ipv6.address_interface_identifier });
      if (config.ipv6.adjust_mss) operations.push({ op: "set_ipv6_adjust_mss", value: config.ipv6.adjust_mss });
      if (config.ipv6.base_reachable_time) operations.push({ op: "set_ipv6_base_reachable_time", value: config.ipv6.base_reachable_time });
      if (config.ipv6.disable_forwarding) operations.push({ op: "set_ipv6_disable_forwarding" });
      if (config.ipv6.dup_addr_detect_transmits) operations.push({ op: "set_ipv6_dup_addr_detect_transmits", value: config.ipv6.dup_addr_detect_transmits });
      if (config.ipv6.source_validation) operations.push({ op: "set_ipv6_source_validation", value: config.ipv6.source_validation });
    }

    // DHCP options
    if (config.dhcp_options) {
      if (config.dhcp_options.client_id) operations.push({ op: "set_dhcp_options_client_id", value: config.dhcp_options.client_id });
      if (config.dhcp_options.default_route_distance) operations.push({ op: "set_dhcp_options_default_route_distance", value: config.dhcp_options.default_route_distance });
      if (config.dhcp_options.host_name) operations.push({ op: "set_dhcp_options_host_name", value: config.dhcp_options.host_name });
      if (config.dhcp_options.mtu) operations.push({ op: "set_dhcp_options_mtu" });
      if (config.dhcp_options.no_default_route) operations.push({ op: "set_dhcp_options_no_default_route" });
      if (config.dhcp_options.reject) operations.push({ op: "set_dhcp_options_reject", value: config.dhcp_options.reject });
      if (config.dhcp_options.user_class) operations.push({ op: "set_dhcp_options_user_class", value: config.dhcp_options.user_class });
      if (config.dhcp_options.vendor_class_id) operations.push({ op: "set_dhcp_options_vendor_class_id", value: config.dhcp_options.vendor_class_id });
    }

    // DHCPv6 options
    if (config.dhcpv6_options) {
      if (config.dhcpv6_options.duid) operations.push({ op: "set_dhcpv6_options_duid", value: config.dhcpv6_options.duid });
      if (config.dhcpv6_options.no_release) operations.push({ op: "set_dhcpv6_options_no_release" });
      if (config.dhcpv6_options.parameters_only) operations.push({ op: "set_dhcpv6_options_parameters_only" });
      if (config.dhcpv6_options.rapid_commit) operations.push({ op: "set_dhcpv6_options_rapid_commit" });
      if (config.dhcpv6_options.temporary) operations.push({ op: "set_dhcpv6_options_temporary" });
      if (config.dhcpv6_options.no_request_dns) operations.push({ op: "set_dhcpv6_options_no_request_dns" });
      if (config.dhcpv6_options.no_request_domain_name) operations.push({ op: "set_dhcpv6_options_no_request_domain_name" });
    }

    // Mirror & Redirect
    if (config.mirror_ingress) operations.push({ op: "set_mirror_ingress", value: config.mirror_ingress });
    if (config.mirror_egress) operations.push({ op: "set_mirror_egress", value: config.mirror_egress });
    if (config.redirect) operations.push({ op: "set_redirect", value: config.redirect });

    return this.batchConfigure(config.name, operations);
  }

  async updateInterface(
    name: string,
    current: MacsecInterface,
    updated: Omit<Partial<MacsecInterface>, "security"> & {
      security_mode?: "mka" | "static";
      security?: Partial<MacsecSecurityConfig> & {
        mka?: Partial<MacsecMkaConfig> | null;
        static?: Partial<MacsecStaticConfig> | null;
      };
    }
  ): Promise<VyOSResponse> {
    const operations: MacsecBatchOperation[] = [];

    // Basic fields
    if (updated.description !== undefined) {
      if (updated.description) operations.push({ op: "set_interface_description", value: updated.description });
      else operations.push({ op: "delete_interface_description" });
    }
    if (updated.source_interface !== undefined) {
      if (updated.source_interface) operations.push({ op: "set_source_interface", value: updated.source_interface });
      else operations.push({ op: "delete_source_interface" });
    }
    if (updated.mtu !== undefined) {
      if (updated.mtu) operations.push({ op: "set_mtu", value: updated.mtu });
      else operations.push({ op: "delete_mtu" });
    }
    if (updated.vrf !== undefined) {
      if (updated.vrf) operations.push({ op: "set_vrf", value: updated.vrf });
      else operations.push({ op: "delete_vrf" });
    }
    if (updated.disabled !== undefined) {
      if (updated.disabled) operations.push({ op: "set_interface_disable" });
      else operations.push({ op: "delete_interface_disable" });
    }

    // Addresses (delete old, set new)
    if (updated.addresses !== undefined) {
      for (const old of current.addresses) {
        operations.push({ op: "delete_interface_address", value: old });
      }
      for (const addr of updated.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }

    // Security
    if (updated.security !== undefined) {
      if (updated.security.cipher !== undefined) {
        if (updated.security.cipher) operations.push({ op: "set_security_cipher", value: updated.security.cipher });
        else operations.push({ op: "delete_security_cipher" });
      }
      if (updated.security.encrypt !== undefined) {
        if (updated.security.encrypt) operations.push({ op: "set_security_encrypt" });
        else operations.push({ op: "delete_security_encrypt" });
      }
      if (updated.security.replay_window !== undefined) {
        if (updated.security.replay_window) operations.push({ op: "set_security_replay_window", value: updated.security.replay_window });
        else operations.push({ op: "delete_security_replay_window" });
      }

      // MKA
      if (updated.security.mka !== undefined) {
        if (updated.security.mka === null) {
          // Clear MKA
          operations.push({ op: "delete_security_mka_cak" });
          operations.push({ op: "delete_security_mka_ckn" });
          operations.push({ op: "delete_security_mka_priority" });
        } else {
          if (updated.security.mka.cak !== undefined) {
            if (updated.security.mka.cak) operations.push({ op: "set_security_mka_cak", value: updated.security.mka.cak });
            else operations.push({ op: "delete_security_mka_cak" });
          }
          if (updated.security.mka.ckn !== undefined) {
            if (updated.security.mka.ckn) operations.push({ op: "set_security_mka_ckn", value: updated.security.mka.ckn });
            else operations.push({ op: "delete_security_mka_ckn" });
          }
          if (updated.security.mka.priority !== undefined) {
            if (updated.security.mka.priority) operations.push({ op: "set_security_mka_priority", value: updated.security.mka.priority });
            else operations.push({ op: "delete_security_mka_priority" });
          }
        }
      }

      // Static
      if (updated.security.static !== undefined) {
        if (updated.security.static === null) {
          // Clear static
          operations.push({ op: "delete_security_static_key" });
          if (current.security?.static?.peers) {
            for (const peer of current.security.static.peers) {
              operations.push({ op: "delete_security_static_peer", value: peer.name });
            }
          }
        } else {
          if (updated.security.static.key !== undefined) {
            if (updated.security.static.key) operations.push({ op: "set_security_static_key", value: updated.security.static.key });
            else operations.push({ op: "delete_security_static_key" });
          }
          if (updated.security.static.peers !== undefined) {
            // Delete old peers
            if (current.security?.static?.peers) {
              for (const peer of current.security.static.peers) {
                operations.push({ op: "delete_security_static_peer", value: peer.name });
              }
            }
            // Set new peers
            for (const peer of updated.security.static.peers) {
              operations.push({ op: "set_security_static_peer", value: peer.name });
              if (peer.key) operations.push({ op: "set_security_static_peer_key", value: `${peer.name}:${peer.key}` });
              if (peer.mac) operations.push({ op: "set_security_static_peer_mac", value: `${peer.name}:${peer.mac}` });
              if (peer.disable) operations.push({ op: "set_security_static_peer_disable", value: peer.name });
            }
          }
        }
      }
    }

    // Mirror
    if (updated.mirror_ingress !== undefined) {
      if (updated.mirror_ingress) operations.push({ op: "set_mirror_ingress", value: updated.mirror_ingress });
      else operations.push({ op: "delete_mirror_ingress" });
    }
    if (updated.mirror_egress !== undefined) {
      if (updated.mirror_egress) operations.push({ op: "set_mirror_egress", value: updated.mirror_egress });
      else operations.push({ op: "delete_mirror_egress" });
    }
    if (updated.redirect !== undefined) {
      if (updated.redirect) operations.push({ op: "set_redirect", value: updated.redirect });
      else operations.push({ op: "delete_redirect" });
    }

    if (operations.length === 0) {
      return { success: true, data: { message: "No changes detected" } };
    }

    return this.batchConfigure(name, operations);
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure(name, [{ op: "delete_interface" }]);
  }
}

export const macsecService = new MacsecService();

File: backend/routers/interfaces/macsec.py
"""
MACsec Interface Configuration Endpoints

All MACsec interface endpoints for VyOS configuration.
MACsec (Media Access Control Security) provides layer-2 encryption
over Ethernet links using IEEE 802.1AE.
"""

import inspect
import logging
from typing import Dict, List, Optional, Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field, ConfigDict
from starlette.concurrency import run_in_threadpool

from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from session_vyos_service import get_session_vyos_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/macsec", tags=["macsec-interface"])


# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Request / Response Models
# ============================================================================


class BatchOperation(BaseModel):
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value (if required)")


class BatchRequest(BaseModel):
    interface: str = Field(..., description="Interface name (e.g., macsec0)")
    operations: List[BatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


class MacsecMkaConfig(BaseModel):
    cak: Optional[str] = None
    ckn: Optional[str] = None
    priority: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecStaticPeer(BaseModel):
    name: str
    key: Optional[str] = None
    mac: Optional[str] = None
    disable: bool = False

    model_config = ConfigDict(populate_by_name=True)


class MacsecStaticConfig(BaseModel):
    key: Optional[str] = None
    peers: List[MacsecStaticPeer] = Field(default_factory=list)

    model_config = ConfigDict(populate_by_name=True)


class MacsecSecurityConfig(BaseModel):
    cipher: Optional[str] = None
    encrypt: bool = False
    replay_window: Optional[str] = None
    mka: Optional[MacsecMkaConfig] = None
    static: Optional[MacsecStaticConfig] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecIpConfig(BaseModel):
    adjust_mss: Optional[str] = None
    arp_cache_timeout: Optional[str] = None
    disable_arp_filter: bool = False
    disable_forwarding: bool = False
    enable_arp_accept: bool = False
    enable_arp_announce: bool = False
    enable_arp_ignore: bool = False
    enable_directed_broadcast: bool = False
    enable_proxy_arp: bool = False
    proxy_arp_pvlan: bool = False
    source_validation: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecIpv6Config(BaseModel):
    accept_dad: Optional[str] = None
    address_autoconf: bool = False
    address_eui64: Optional[str] = None
    address_no_default_link_local: bool = False
    address_interface_identifier: Optional[str] = None
    adjust_mss: Optional[str] = None
    base_reachable_time: Optional[str] = None
    disable_forwarding: bool = False
    dup_addr_detect_transmits: Optional[str] = None
    source_validation: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecDhcpOptions(BaseModel):
    client_id: Optional[str] = None
    default_route_distance: Optional[str] = None
    host_name: Optional[str] = None
    mtu: bool = False
    no_default_route: bool = False
    reject: Optional[str] = None
    user_class: Optional[str] = None
    vendor_class_id: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecDhcpv6PdInterface(BaseModel):
    address: Optional[str] = None
    sla_id: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecDhcpv6Pd(BaseModel):
    length: Optional[str] = None
    interfaces: Dict[str, MacsecDhcpv6PdInterface] = Field(default_factory=dict)

    model_config = ConfigDict(populate_by_name=True)


class MacsecDhcpv6Options(BaseModel):
    duid: Optional[str] = None
    no_release: bool = False
    no_request_dns: bool = False
    no_request_domain_name: bool = False
    parameters_only: bool = False
    rapid_commit: bool = False
    temporary: bool = False
    pd: Optional[Dict[str, MacsecDhcpv6Pd]] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecInterfaceConfig(BaseModel):
    name: str
    type: str
    addresses: List[str] = Field(default_factory=list)
    description: Optional[str] = None
    disabled: bool = False
    mtu: Optional[str] = None
    source_interface: Optional[str] = None
    vrf: Optional[str] = None
    security: Optional[MacsecSecurityConfig] = None
    ip: Optional[MacsecIpConfig] = None
    ipv6: Optional[MacsecIpv6Config] = None
    dhcp_options: Optional[MacsecDhcpOptions] = None
    dhcpv6_options: Optional[MacsecDhcpv6Options] = None
    mirror_ingress: Optional[str] = None
    mirror_egress: Optional[str] = None
    redirect: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class MacsecInterfacesConfigResponse(BaseModel):
    interfaces: List[MacsecInterfaceConfig] = Field(default_factory=list)
    total: int = 0


# ============================================================================
# Endpoints
# ============================================================================


@router.get("/capabilities")
async def get_capabilities(request: Request) -> Dict[str, Any]:
    """Return version-aware feature capabilities for MACsec interfaces."""
    await require_read_permission(request, FeatureGroup.INTERFACES)
    service = get_session_vyos_service(request)
    from vyos_builders.interfaces.macsec import MacsecInterfaceBuilderMixin
    builder = MacsecInterfaceBuilderMixin(version=service.get_version())
    return builder.get_capabilities()


@router.get("/config", response_model=MacsecInterfacesConfigResponse)
async def get_config(http_request: Request, refresh: bool = False) -> MacsecInterfacesConfigResponse:
    """Get all MACsec interface configurations from VyOS."""
    await require_read_permission(http_request, FeatureGroup.INTERFACES)
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh)
        raw_config = full_config.get("interfaces", {}).get("macsec", {})

        from vyos_mappers.interfaces.macsec_versions import get_macsec_mapper
        mapper = get_macsec_mapper(service.get_version())
        parsed = mapper.parse_interfaces_of_type(raw_config)
        return MacsecInterfacesConfigResponse(**parsed)
    except Exception:
        logger.exception("Unhandled error in get_config")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/batch", response_model=VyOSResponse)
async def batch_configure(http_request: Request, request: BatchRequest) -> VyOSResponse:
    """
    Configure a MACsec interface using batch operations.

    **Supported operations:**
    | Operation | Value | Description |
    |-----------|-------|-------------|
    | `set_interface_description` | Yes | Set description |
    | `delete_interface_description` | No | Remove description |
    | `set_interface_address` | Yes | Add IP address (CIDR) |
    | `delete_interface_address` | Yes | Remove IP address |
    | `set_interface_disable` | No | Disable interface |
    | `delete_interface_disable` | No | Enable interface |
    | `set_mtu` | Yes | Set MTU (68-16000) |
    | `delete_mtu` | No | Reset MTU to default |
    | `set_source_interface` | Yes | Set source ethernet interface |
    | `delete_source_interface` | No | Remove source interface |
    | `set_vrf` | Yes | Assign to VRF |
    | `delete_vrf` | No | Remove VRF assignment |
    | `set_security_cipher` | Yes | Set cipher (gcm-aes-128/gcm-aes-256) |
    | `delete_security_cipher` | No | Remove cipher |
    | `set_security_encrypt` | No | Enable encryption |
    | `delete_security_encrypt` | No | Disable encryption |
    | `set_security_replay_window` | Yes | Set replay window (0-4294967295) |
    | `delete_security_replay_window` | No | Remove replay window |
    | `set_security_mka_cak` | Yes | Set MKA CAK (hex) |
    | `delete_security_mka_cak` | No | Remove MKA CAK |
    | `set_security_mka_ckn` | Yes | Set MKA CKN (hex) |
    | `delete_security_mka_ckn` | No | Remove MKA CKN |
    | `set_security_mka_priority` | Yes | Set MKA priority (0-255) |
    | `delete_security_mka_priority` | No | Remove MKA priority |
    | `set_security_static_key` | Yes | Set static key (hex) |
    | `delete_security_static_key` | No | Remove static key |
    | `set_security_static_peer` | Yes | Create static peer |
    | `delete_security_static_peer` | Yes | Delete static peer |
    | `set_security_static_peer_disable` | Yes | Disable static peer |
    | `delete_security_static_peer_disable` | Yes | Enable static peer |
    | `set_security_static_peer_key` | Yes | Set peer key (peer:key) |
    | `delete_security_static_peer_key` | Yes | Remove peer key |
    | `set_security_static_peer_mac` | Yes | Set peer MAC (peer:mac) |
    | `delete_security_static_peer_mac` | Yes | Remove peer MAC |
    | `delete_interface` | No | Delete entire interface |
    """
    await require_write_permission(http_request, FeatureGroup.INTERFACES)

    try:
        service = get_session_vyos_service(http_request)
        batch = service.create_macsec_batch()

        for op in request.operations:
            if op.op in batch._INTERNAL_BUILDER_METHODS:
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation '{op.op}' is not a valid interface operation",
                )

            method = getattr(batch, op.op, None)
            if method is None:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported operation: {op.op}",
                )

            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 1:
                method(request.interface)
            elif len(params) == 2:
                if op.value is None:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires a value",
                    )
                method(request.interface, op.value)
            elif len(params) == 3:
                if op.value is None:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires a value",
                    )
                parts = op.value.split(":", 1)
                if len(parts) != 2:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires value in 'param1:param2' format",
                    )
                method(request.interface, parts[0], parts[1])
            elif len(params) == 4:
                if op.value is None:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires a value",
                    )
                parts = op.value.split(":", 2)
                if len(parts) != 3:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires value in 'param1:param2:param3' format",
                    )
                method(request.interface, parts[0], parts[1], parts[2])
            elif len(params) == 5:
                if op.value is None:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires a value",
                    )
                parts = op.value.split(":", 3)
                if len(parts) != 4:
                    raise HTTPException(
                        status_code=400,
                        detail=f"Operation '{op.op}' requires value in 'param1:param2:param3:param4' format",
                    )
                method(request.interface, parts[0], parts[1], parts[2], parts[3])
            else:
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation '{op.op}' has unexpected signature",
                )

        response = service.execute_batch(batch)
        return VyOSResponse(
            success=response.status == 200,
            data=response.result if isinstance(response.result, dict) else None,
            error=response.error if response.error else None,
        )
    except HTTPException:
        raise
    except Exception:
        logger.exception("Unhandled error in batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")
